rstar/src/algorithm/iterators.rs

```diff
@@ -104,22 +104,21 @@ where
 {
     type Item = &'a mut T;
     fn next(&mut self) -> Option<&'a mut T> {
-        let func = &self.func;
-        if let Some(next) = self.current_nodes.pop() {
+        while let Some(next) = self.current_nodes.pop() {
             match next {
-                RTreeNode::Leaf(ref mut t) => Some(t),
+                RTreeNode::Leaf(ref mut t) => {
+                    if self.func.should_unpack_leaf(t) {
+                        return Some(t);
+                    }
+                }
                 RTreeNode::Parent(ref mut data) => {
-                    self.current_nodes.extend(
-                        data.children
-                            .iter_mut()
-                            .filter(|c| func.should_unpack_parent(&c.envelope())),
```

Comment:
The filtering here isn't done anymore, so the leaf `envelope` isn't checked if `should_unpack_leaf` doesn't do it. This means `current_nodes` might grow more than it would when checking is eager.

This could be added back, but should be symmetrical with `SelectionIterator`. If that happens `should_unpack_leaf` should probably have docs to say its envelope is already checked, and the trait could default to `true`:
```rust
fn should_unpack_leaf(&self, leaf: &T) -> bool {
    return true;
}
```